### PR TITLE
90 average calibration

### PR DESCRIPTION
**Pull request by pasq-cat**
This pull request add the following things:

- Add a method to compute the average level of calibration for both classification and regression tasks, using the method proposed by kuleshov in his paper https://arxiv.org/pdf/1807.00263
-  add unit tests and docstrings for the functions implemented.
- update and expand the documentation on the website.

### CODE REVIEW

**Comment by Rockdeldiablo** [comment]
@pat-alt ok this one passes the tests but i don't want it to be merged yet for some reasons:
1) there is a small function that is still missing 
2) due to the Stochastic nature of sampling there is a small chance that some test fails from time to time
3) after i write the docs and talk to the guys in calibration.jl maybe everything will be clearer

so i think it's best if we leave this one to late august.

**Comment by pat-alt** [comment]
OK, sounds good to me!

File: .github/workflows/CI.yml
Comment (pat-alt) [inline .github/workflows/CI.yml:1]: Is this change necessary? I think the existing approach should use the latest release?

File: docs/src/tutorials/calibration.qmd
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:1]: I think this qmd needs to be re-rendered one more time?

File: src/calibration_functions.jl [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+"""
+function sharpness_classification(y_binary, distributions::Vector{Bernoulli{Float64}})
+    mean_class_one = mean(mean.(distributions[findall(y_binary .== 1)]))
+    mean_class_zero = mean(1 .- mean.(distributions[findall(y_binary .== 0)]))
+    return mean_class_one, mean_class_zero
+end
+
+@doc raw""" 
+    empirical_frequency_classification(y_binary, distributions::Distributions.Bernoulli)
```
Comment (pat-alt) [inline src/calibration_functions.jl:229]: Are these docstrings rendering correctly in the REPL when you use help mode?

**Review by pat-alt — COMMENTED** [review]
Just a few minor things (see individual comments)

File: .github/workflows/CI.yml
Comment (pasq-cat) [reply .github/workflows/CI.yml:1]: uhm i thought i removed julia version 1.0 and left 1.10. how do i fix this? replace 1.9 with 1.10?

File: docs/src/tutorials/calibration.qmd
Comment (pasq-cat) [reply docs/src/tutorials/calibration.qmd:1]: yes they are not rendered yet because i am still working on them. anyway, now that we are at it, a question? whenever i type quarto render tutorials/etc i end up with a lot of files that they are not used. in particular i get 2 types of markdown files and other stuff that i have to delete manually. is there a way to avoid this issue?

File: src/calibration_functions.jl (hunk 1 above)
Comment (pasq-cat) [reply src/calibration_functions.jl:229]: i will check one for one. at the moment i am working on the multi.qmd file. i am failing to make to join the plots in a 2,2 grid. I will upload as it is now , so that you can see.

Comment (pasq-cat) [reply src/calibration_functions.jl:229]: btw i was considering deleting the original functions for sampled distributions and leaving only those for normal distributions since laplaceredux only works for this kind of distr. i will submit the general ones to the guys at calibration.jl @pat-alt

File: test/calibration.jl [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+    # Test 2: Check the function with a known input
+    distributions = [
+        Distributions.Normal.(1, 0.01),
+        Distributions.Normal.(1, 2),
+        Distributions.Normal.(1, 3),
+    ]
+    mean_variance = mean(map(var, distributions))
+    sharpness = sharpness_regression(distributions)
+    @test sharpness ≈ mean_variance
```
Comment (pat-alt) [inline test/calibration.jl:185]: Can adjust the tolerance here: 

```julia
julia> @test π ≈ 3.14 atol=0.01
```

File: .github/workflows/CI.yml
Comment (pat-alt) [reply .github/workflows/CI.yml:1]: Actually, nvm, let's keep as is now

File: docs/src/tutorials/calibration.qmd
Comment (pat-alt) [reply docs/src/tutorials/calibration.qmd:1]: As discussed, add this to _quarto.yml

```yaml
render:
    - "!docs/src/*.md"
```

File: _quarto.yml [hunk 3]
```diff
@@ -2,6 +2,9 @@ project:
   title: "LaplaceRedux.jl"
   execute-dir: project
 
+render:
```
Comment (pasq-cat) [inline _quarto.yml:5]: @pat-alt  i tried placing this like piece of code both here and in  /docs/_metadata.yml but it doesn't work. if i run quarto render . from inside the /tutorials folder i get 44 new files.

Comment (pat-alt) [reply _quarto.yml:5]: I think this needs to be indented to fall under the project settings:

```yaml
project:
  title: "{{{PKG}}}.jl"
  execute-dir: project
  render:
    - "!docs/src/*.md"
```

Comment (pasq-cat) [reply _quarto.yml:5]: @pat-alt no, it stops working completely if i do this. btw how do you post code blocks in the comments?

Comment (pat-alt) [reply _quarto.yml:5]: Just markdown syntax https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

Comment (pat-alt) [reply _quarto.yml:5]: What do you mean by "it stops working completely"?

Comment (pasq-cat) [reply _quarto.yml:5]: nvm found the solution

Comment (pasq-cat) [reply _quarto.yml:5]: @pat-alt the documentation says that when you use the prefix ! "you need to start by specifying everything you do want to render. For example: project:
  render:
    - "*.qmd"
    - "!ignored.qmd"
    - "!ignored-dir/"

Comment (pat-alt) [reply _quarto.yml]: You're right, thanks!

Comment (pasq-cat) [reply _quarto.yml]: @pat-alt  is there a way to preview the website? i would like to check if everything is okay but it doesn't seems to be a command. Seems like i am missing a step... once i render the .qmd files, how do i rerender the html part?

Comment (pasq-cat) [reply _quarto.yml]: > This is handled by Documenter.jl, simply run `julia --project=docs docs/make.jl` in the command line. This will render to the `docs/build` directory. In there, simply click on the `index.html` file, which will open the (local version of the) website in your browser

is it always this complicate or you just set it up like this xD?

Comment (pasq-cat) [reply _quarto.yml]: @pat-alt i don't know why but some latex  equations do not get rendered in the html version, although quarto shows they are correctly written in the little square preview box on the left side of the screen.

**Comment by pat-alt** [comment]
Let me know when you're happy for this to be merged, then I'll review one more time @pasq-cat

**Comment by pasq-cat** [comment]
> Let me know when you're happy for this to be merged, then I'll review one more time @pasq-cat

try  this version pls. i had to add seed to the arguments of the functions in data.jl. see if it's ok for you.

File: docs/src/tutorials/calibration.qmd [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+``` @meta
+CurrentModule = LaplaceRedux
+```
+
+# Uncertainty Calibration
+
+# The issue of calibrated uncertainty distributions
+
+
+Bayesian methods offer a general framework for quantifying uncertainty. However, due to model misspecification and the use of approximate inference techniques,uncertainty estimates are often inaccurate: for example, a 90% credible interval may not contain the true outcome 90% of the time, in such cases the model is said to be miscalibrated.
```
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:12]: "inference techniques,uncertainty" missing space after comma

File: docs/src/tutorials/calibration.qmd [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+```
+
+# Uncertainty Calibration
+
+# The issue of calibrated uncertainty distributions
+
+
+Bayesian methods offer a general framework for quantifying uncertainty. However, due to model misspecification and the use of approximate inference techniques,uncertainty estimates are often inaccurate: for example, a 90% credible interval may not contain the true outcome 90% of the time, in such cases the model is said to be miscalibrated. 
+
+ This problem arises due to the limitations of the model itself: a predictor may not be sufficiently expressive to assign the right probability to every credible interval, just as it may not be able to always assign the right label to a datapoint.\ Miscalibrated credible intervals reduce the trustworthyness of the  forecaster because they lead to a false sense of precision  and either overconfidence or underconfidence in the results.
```
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:14]: Note that the "\" character is ignored. If you want a line break, you can just literally add a line break in the `qmd` file

Comment (pat-alt) [reply docs/src/tutorials/calibration.qmd:14]: "trustworthiness"

File: docs/src/tutorials/calibration.qmd [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+A forecaster is said to be perfectly calibrated if a 90% credible interval contains the true outcome approximately 90% of the time. Perfect calibration however cannot be achieved with limited data, because with limited data comes inherent statistical fluctuations that can cause the estimated credible intervals to deviate from the ideal coverage probability. Furthermore, a finite sample of collected data points cannot eliminate completely the influence of the possible misjudged priors probabilities.
```
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:16]: "prior probabilities"? the extra "s" is not needed

File: docs/src/tutorials/calibration.qmd [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+On top of these issues, which stem directly from Bayes' theorem, with Bayesian neural network there is also the problems introduced by the approximate inference method adopted to compute the posterior distribution of the weights.\  To introduce the concept of Average Calibration and the Calibration Plots, we will follow closely the paper [Accurate Uncertainties for Deep Learning Using Calibrated Regression](https://arxiv.org/abs/1807.00263), written by Volodymyr Kuleshov, Nathan Fenner and Stefano Ermon, although with some small differences. We will hightlight these differences in the following paragraphs whenever they appear. We present here the theoretical basis necessary to understand the issue of calibration and we refer to the tutorials for the coding examples.
```
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:18]: "Bayesian neural networks" plural

File: docs/src/tutorials/calibration.qmd [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+$$\frac{\sum_{t=1}^T \mathbb{1} \{ y_t \leq F_t^{-1}(p)  \} }{T} \rightarrow p \quad \text{for all}\quad  p \in [0,1]$$
+
+as $T \rightarrow \infty$.\  In other words, the empirical and the predicted CDFs should match as the dataset size goes to infinity.
+Perfect Calibration is a sufficient condition for average calibration, the opposite however is not necessarily true: a model can be average calibrated but not perfectly calibrated.\
+From now on when we talk about calibration we will implicitly talk about average calibration rather than perfect calibration. 
+
+
+### Sharpness 
+Calibration by itself is not sufficient to produce a useful forecast. For example, it is easy to see that if we use for the forecast the marginal distribution $F(y) = \mathbb{P}(Y ≤ y)$, without considering the input feature $X$, the forecast will be calibrated but still not accurate. In order to be useful, forecasts must also be sharp, which,in a regression context, means that the confidence intervals should all be as tight as possible around a single value. More formally, we want the variance $var(F_t)$ of the random variable whose CDF is $F_t$ to be small.\
```
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:35]: "if we use the marginal distribution $F(y) = \mathbb{P}(Y ≤ y)$ for the forecast without ..."

Comment (pat-alt) [reply docs/src/tutorials/calibration.qmd:35]: "which,in a regression context, means" -> "which (in a regression context) means ..."

File: docs/src/tutorials/calibration.qmd [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+As  a sharpness score of the forecaster, Kuleshov et al. proposed the average predicted variance
+
+$$sharpness(F_{1} ,\dots, F_T) = \frac{1}{T} \sum_{t=1}^T var{F_t}$$ \
+the smaller the sharpness, the tighter will be the confidence intervals on average.
+
+### Calibration Plots 
+
+To check the level of calibration, Kuleshov et al. proposed a calibration plot that display the true frequency of points in each confidence interval relative to the predicted fraction of points in that interval.\
```
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:43]: "displays"

File: docs/src/tutorials/calibration.qmd [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+More formally, we choose $m$ confidence levels $0 ≤ p_1 < p_2 < . . . < p_m ≤ 1$; for each threshold $p_j$ , and compute the empirical frequency
+
+$$\hat{p}_j = \frac{|\{ y_t|F_t(y_t) \leq p_t, t= 1,2,\dots,T    \} |}{T}.$$
+
+To visualize the level of average calibration, we plot $\{(p_j,\hat{p_j}) \}_{j=1}^M$; A forecaster that is calibrated  will correspond a straight line on the plot that goes from $\{0,0\}$ to $\{1,1\}$ . \
```
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:48]: "correspond to"

File: docs/src/tutorials/calibration.qmd [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+As a measure of the level of miscalibration of the forecaster, differently from what it is suggested in the original paper, we a measure  the area between the bisector of the plot and the line produced by the forecaster.
```
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:49]: "we measure"

Might need to elaborate a bit here on why we do this differently and how exactly the area is computed/measured.

File: docs/src/tutorials/calibration.qmd [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+ 
+### Post-training calibration
+As we  have said previously, uncertainty estimates obtained by deep BNNs tends to be miscalibrated. We introduced the support to a post-training technique for regression problems presented in  [Recalibration of Aleatoric and Epistemic Regression Uncertainty in Medical Imaging](https://arxiv.org/abs/2104.12376)
```
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:53]: "tend"

File: docs/src/tutorials/calibration.qmd [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+by Max-Heinrich Laves, Sontje Ihler, Jacob F. Fast, Lüder A. Kahrs and Tobias Ortmaier and usually referred to as sigma-scaling. Using a Gaussian model, the technique consist in scaling the predicted standard deviation $\sigma$ with a scalar value $s$ to recalibrate the probability density function
+
+$$p(y|x; \hat{y}(x), \hat{σ}^2(x)) = \mathbb{N}( y; \hat{y}(x),(s  \cdot  \hat{σ}(x))^2 ).$$ 
+
+This results in the following minimization objective:
+
+$$L_G(s) = m \log(s) + \frac{1}{2}s^{−2} \sum_{i=1}^m (\hat{σ}^{(i)}_θ)^{−2} || y^{(i)} − \hat{y}_{\theta}^{(i)}||^2.$$
+
+In general, this equation can be optimized  respect to s with fixed values for the parameters $\theta$ using gradient descent in a second phase over a separate calibration set.  However, for the case of a gaussian distribution, the analytical solution is known  and takes the closed form
```
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:62]: "respect to $s$ with"

File: docs/src/tutorials/calibration.qmd [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+$$s = \pm \sqrt{\frac{1}{m} \sum_{i=1}^m (\hat{σ}^{(i)}_θ)^{−2} || y^{(i)} − \hat{y}_{\theta}^{(i)}||^2}.$$
+
+Once the scalar $s$ is computed, all we have to do to obtain better calibrated predictions is to multiply the predicted standard deviation with the scalar.
+
```
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:67]: Might need some explanation here as to how this differs from Empirical Bayes (currently use to tune the prior, including observational noise in the regression case).

File: docs/src/tutorials/calibration.qmd [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+## Calibration Plots
+ Given a dataset ${(x_t, y_t)}^T_t=1$, let $p_t = H(x_t) ∈ [0, 1]$ be the forecasted probability. We group the $p_t$ into intervals $I_j$ for
+$j = 1, 2, ..., m$ that form a partition of $[0, 1]$. \
+A calibration curve plots the predicted average $ p_j = T^{−1}_j \sum_{t:p_t∈ I_j} p_t $ in each interval $I_j$ against the observed empirical average 
+$$p_j = T^{−1}_j \sum_{t:p_t ∈ I_j}y_t,$$ where $T_j = |{t : p_t ∈ I_j}|$. Perfect calibration corresponds once again to a straight line.
+
+
+
+### multiclass case
```
Comment (pat-alt) [inline docs/src/tutorials/calibration.qmd:93]: `### Multiclass Case`

File: docs/src/tutorials/logit.qmd [hunk 16]
```diff
@@ -74,10 +102,33 @@ optimize_prior!(la; verbose=true, n_steps=500)
...
+ predicted_distributions= predict(la, X_test,ret_distr=true)
+```
+
+then we plot the calibration plot
+```{julia} 
+#| output: true
+Calibration_Plot(la,ys_test,vec(predicted_distributions);n_bins = 10)
+``` 
+
+as we can see from the plot, although extremely accurate, the neural network doesn't seem to be calibrated well. This is however an effect of the extreme accuracy reached by the neural network which causes the lack of predictions with  high uncertainty (low certainty). We can see this by looking at the level of sharpness for the two classes which are extremely close to 1, indicating the high level of trust that the neural network has in the predictions.
```
Comment (pat-alt) [inline docs/src/tutorials/logit.qmd:127]: "does not"

Comment (pat-alt) [reply docs/src/tutorials/logit.qmd:127]: "This is, however, ..."

File: docs/src/tutorials/multi.qmd [hunk 17]
```diff
@@ -67,26 +100,67 @@ fit!(la, data)
...
     push!(plt_list, plt)
 end
 plot(plt_list...)
-```
\ No newline at end of file
+```
+
+## Calibration Plots
+
+In the case of multiclass classification tasks, we cannot plot directly the calibration plots since they can only be used in the binary classification case. However, we can use them to plot the calibration of the predictions for 1 class against all the others. To do so, we first have to collect the predicted categorical distributions
```
Comment (pat-alt) [inline docs/src/tutorials/multi.qmd:133]: "we cannot plot directly the calibration plots" -> "we cannot plot the calibration plots directly"

File: docs/src/tutorials/multi.qmd [hunk 18]
```diff
@@ -67,26 +100,67 @@ fit!(la, data)
...
+```
+
+then we transform the categorical distributions into Bernoulli distributions by taking only the probability of the class of interest, for example the third one. 
+```{julia}
+#| output: true
+using Distributions
+bernoulli_distributions = [Bernoulli(p.p[3]) for p in vec(predicted_distributions)]
+```
+
+Now we can use the calibration Plot to see the level of calibration of the neural network
```
Comment (pat-alt) [inline docs/src/tutorials/multi.qmd:147]: use actual function name

File: docs/src/tutorials/multi.qmd [hunk 19]
```diff
@@ -67,26 +100,67 @@ fit!(la, data)
...
+```{julia}
+#| output: true
+plt = Calibration_Plot(la,hcat(y_onehot_test...)[3,:],bernoulli_distributions;n_bins = 10);
+
+```
+
+The plot is peaked around 0.7, and This may be due to various reasons.
+
+A possible reason is that class 3 is relatively easy for the model to identify from the other classes. The network is able to  correctly identify examples belonging to class 3, although he remains a bit underconfident in its predictions.
```
Comment (pat-alt) [inline docs/src/tutorials/multi.qmd:156]: "although it remains"

File: docs/src/tutorials/multi.qmd [hunk 20]
```diff
@@ -67,26 +100,67 @@ fit!(la, data)
...
+plt = Calibration_Plot(la,hcat(y_onehot_test...)[3,:],bernoulli_distributions;n_bins = 10);
+
+```
+
+The plot is peaked around 0.7, and This may be due to various reasons.
+
+A possible reason is that class 3 is relatively easy for the model to identify from the other classes. The network is able to  correctly identify examples belonging to class 3, although he remains a bit underconfident in its predictions. 
+Another reason for the peak however may be the lack of cases where the predicted probability is lower (e.g., around 0.5), which could indicate that the network  has not encountered ambiguous or difficult-to-classify examples for such class.  This once again might be because either class 3 has distinct features that the model can easily learn, leading to fewer uncertain predictions, or is a consequence of the limited dataset. 
+
+ We can measure how sharp is the neural network by computing the sharpness score
```
Comment (pat-alt) [inline docs/src/tutorials/multi.qmd:159]: "We can measure how sharp is the neural network" -> "We can measure how sharp the neural network is"

**Review by pat-alt — COMMENTED** [review]
Nice work!

Small comments mostly about typos/grammar. 

One thing I noticed is the `Calibration_plot` function name. Can we just make this `calibration_plot` please? I guess that's in TaijaPlotting though

File: docs/src/tutorials/calibration.qmd (hunk 11 above)
Comment (pasq-cat) [reply docs/src/tutorials/calibration.qmd:49]: it's not really different, more like a simplification  pic . he uses the rectangular rule with variable weights, i use the trapezoidal rule.

File: test/calibration.jl (hunk 2 above)
Comment (pasq-cat) [reply test/calibration.jl:25]: no need anymore, it was a bug i think.
